Support cache coverage

shimoncohen opened the pull request:
Add coverage support for caches #690.
This also adds support for #304 and #659.

weskamm commented:
This sounds promising, thanks so far. Could you please add some tests?

shimoncohen commented:
Sure, I will work on it!

shimoncohen commented:
@weskamm Added tests, could you please have a look?

weskamm commented:
A coverage can also be a (multi)polygon, see https://mapproxy.github.io/mapproxy/coverages.html#coverages

Your code looks like that it only handles / respects extent / bbox, but not real polygons. Do you think you could extent it to also handle that? If not, it should be stated very clearly that this feature is limited to the extent of a geometry only.

shimoncohen commented:
Thank you very much for bringing this to my attention. I checked to make sure, and this seems to support other types of coverages as expected.
I added cases for `intersection` and `(multi)polygon` to the cache coverage tests.
Does this seem to cover this issue?

weskamm commented:
Well, if i define a coverage for a cache, which e.g. is a polygon of france, i would expect the cache to not have any tiles in it that are outside of france. A bbox or extent of france also covers parts of switzerland. If the implementation only handles the extent / bbox, there will be much more tiles in the cache than expected.
I am not 100% sure here but i think we currently have that issue described. Could you check for this case and add a test which checks if no tiles are chaced outside of a real polygon?

weskamm commented:
Could you please explain your exact use case for this pull request? After think about it for a while I am not sure if your problem could already be solved with exisiting solutions. If, for example, you have a directory with exisiting tiles which you want to filter with a coverage, you could also define the exisiting tiles as a source and use the coverage option on the source. Sorry for asking this question so late, but i am not sure if this is really necessary.

shimoncohen commented:
Gladly! As I mentioned in #691, from what I understand, Mapproxy's support for `sources` is only for some servers.
In our case, we want to serve data we have in `cache` types such as a `geopackage` or `s3` (In general, defining a cache with no sources as Mapproxy allows).
The goal of this PR is to add the ability to be able to define a `coverage` for each `cache` so it will be handled the same way it is currently when defined in the `sources` section.

If we would list the "sub-goals" in case you add a coverage for a `cache`:
* Extent in demo is correctly set so exact data could be displayed (not whole world extent which is default)
* Capabilities report exact BBOX for each layer and not whole world which is default
* Data out of defined coverage is not saved / displayed
* In case of `geopackage` the extent is set to be as coverage BBOX

weskamm commented:
Ok, i think i got a better understanding of your usecase:
You have an already existing cache filled with data. You want to filter that cache in order to not deliver the whole content, but a subset instead. You also want the capabilities to show the custom defined bbox.

Question:
As you can / need to define a grid for your cache, why dont you just define the bbox on the grid? It should solve your problems altogether, or am i missing something?

The problems i currently see with your solution:
  * You seem to only handle the bbox case, which is enough to calculate an extent / bbox for capabilities documents. But as mentioned before, coverages can be polygons. If you only need a bbox in the end, why not introduce a `bbox` parameter on the cache (if the above suggestion does not work). That would avoid confusion because people defining a polygon as coverage may expect a different behaviour than always clipping to its extent.
  * People may have an empty cache, connected with a source. And they want to seed that source. With this change, they can now configure a coverage on the cache and source. This is a somewhat conflicting situation (defining a larger area on the cache than on the source -> makes no sense, will never git filled completely; defining a smaller area on the cache than on the source -> does not make sense, use the source coverage directly). When the source has no coverage restrictions set but the cache, mapproxy will request tiles which are then not allowed to save in the cache, as the cache coverage may be smaller, which would produce undesired traffic and behaviour.
  
Please rethink the idea if a bbox is enough for your usecase and if you can just use the grid for that. If you have a valid usecase together with mapproxy configuration and data, i happily search for further solutions / ideas if you post it here.

shimoncohen commented:
You understand the core use case, but this is not exactly what is implemented. This addition supports all types of coverage (like with sources), not just BBOX. And the use case is also ment to support any kind of `coverage`, which is why defining a `grid` `BBOX` is not enough.

In addition, I added an example for using a `geojson` datasource as coverage to the tests. Do you see a specific case that does not work with my implementation? The only place that a `BBOX` is specifically used is in the `capabilities` and `demo` because they use `BBOX` by definition.

After your notes I also added the handling of requests outside of coverage so they won't be stored or requested from cache (or the source it is defined over). From my checks it works with the combination of coverages on `source` and `cache`.
I haven't added tests for this very important use case yet because I want to make sure you approve of the solution.

weskamm commented on mapproxy/cache/base.py:
```diff
@@ -42,6 +42,9 @@ class TileCacheBase(object):
     supports_timestamp = True
 
+    def __init__(self, coverage=None) -> None:
+        self.coverage = coverage
```
Where is the coverage used in this class? I cannot see it, can you give me a point where i can see where the coverage is handled for the different caches?

shimoncohen replied:
Sure!

The `cache` coverage is accessed in the `loader` class which calculates each caches extent in the `caches` function before return (line 1668):
https://github.com/mapproxy/mapproxy/pull/691/files#diff-7557b8794cb455eb559651f8fa4730d2a8a55d2c13bd188d730eecb4ed10e863R1668

It is also accessed in the `tile` class under cache directory in the `load_tile_coords` function in order to remove tiles that are not in the `coverage` (line 156):
https://github.com/mapproxy/mapproxy/pull/691/files#diff-cc6dc27bf169f009bbbd33027f122d17a5cac2e3f4224db2b58d3869b59673d9R156

weskamm replied:
Thanks

weskamm commented:
Ok i think i am fine with the idea now, thanks so far. Please go on with the test you wanted to add.

shimoncohen commented:
@weskamm Added tests.

weskamm commented:
Thanks, this looks good. Could you finally enhance the documentation and add some information about the usecases when this should be used and the possible conflicts when misconfiguring source and chache coverage at the same time?

shimoncohen commented:
@weskamm Added note about using `coverages` with both `source` and it's defined `cache`.

shimoncohen commented:
Currently in this PR, when there is a `cache` defined over a `source`, the `coverage` used for capabilities and demo will be of the `source`.
I am wondering if this should be changed such that the `cache` coverage will be used, because it mainly defines the layer. Should this be changed?

Note: When using a `cache` with no source or `source` with no `cache` it is used correctly.
